lib/route.js

```diff
@@ -320,19 +328,21 @@ function buildRouting (options) {
           }
         })
 
-        const { exposeHeadRoute } = opts
-        const hasRouteExposeHeadRouteFlag = exposeHeadRoute != null
-        const shouldExposeHead = hasRouteExposeHeadRouteFlag ? exposeHeadRoute : globalExposeHeadRoutes
-
-        if (shouldExposeHead && options.method === 'GET' && !headRouteExists) {
-          const onSendHandlers = parseHeadOnSendHandlers(opts.onSend)
-          prepareRoute.call(this, { method: 'HEAD', url: path, options: { ...opts, onSend: onSendHandlers } })
-        } else if (headRouteExists && exposeHeadRoute) {
-          warning.emit('FSTDEP007')
-        }
-
         done(notHandledErr)
       })
+
+      // register head route in sync
+      // we must place it after the `this.after`
+      const { exposeHeadRoute } = opts
+      const hasRouteExposeHeadRouteFlag = exposeHeadRoute != null
+      const shouldExposeHead = hasRouteExposeHeadRouteFlag ? exposeHeadRoute : globalExposeHeadRoutes
+
+      if (shouldExposeHead && options.method === 'GET' && !hasHEADHandler) {
+        const onSendHandlers = parseHeadOnSendHandlers(opts.onSend)
+        prepareRoute.call(this, { method: 'HEAD', url: path, options: { ...opts, onSend: onSendHandlers }, isFastify: true })
+      } else if (hasHEADHandler && exposeHeadRoute) {
+        warning.emit('FSTDEP007')
+      }
```

Comment:
Could we just do this and keep emitting `onRoute` for `HEAD` routes too?

Reply:
No, we need to remove the `fastify` one when it is needed otherwise the user is force to added `HEAD` route before `GET` route.

Reply:
> No, we need to remove the fastify one when it is needed otherwise user is force to add the HEAD route before GET route.

I thought this was already the case. I had to fix quite a few regressions due to this.

Reply:
> I thought this was already the case.

Actually, no. It is force to create route in sync, but not restricted the order.

For example, 
```js
import Fastify from 'fastify'

{
  // order is not restricted
  const fastify = Fastify()
  
  fastify.get('/', async () => '')
  fastify.head('/', async () => '') // it works
}

{
  // order is not restricted
  const fastify = Fastify()
  
  fastify.get('/', async () => '')
  fastify.after(() => {
    fastify.head('/', async () => '') // it just ignored without any warning.
  })
}
```